Documentation for All-In-One docker image (#1487)
dotse/zonemaster: docs/public/using/gui/all-in-one.md

# Using the All-In-One docker image

## Table of contents
* [All-In-One](#all-in-one)
* [Using Docker](#using-docker)
* [Invoking the command line tool using AIO image](#invoking-the-command-line-tool-using-aio-image)
* [Docker tips](#docker-tips)

## All-In-One

The `zonemaster-all-in-one` docker image provides an All-In-One solution containing Zonemaster-CLI, Zonemaster-Backend, and Zonemaster-GUI.

The `zonemaster-all-in-one` image can be run from any computer that has Docker installed.

### Using Docker

To run Zonemaster via the All-In-One Docker image you have to make sure that Docker is installed
and running.
* Instructions for installation are found on Docker [Get started] page.
* Run the command `docker ps` on the command line to verify that you can run
  Docker on the computer.

When Docker has been correctly installed, no more installation is needed to run
`zonemaster-cli`. Just follow the examples below.

There is a limitation in Docker regarding IPv6. Unless IPv6 has been enabled in
the Docker daemon, there is no support for IPv6. To avoid meaningless errors,

> use `--no-ipv6` if there is no IPv6 support. Also see section "[IPv6 support]".

Reviewer: ```suggestion
use `--no-ipv6` if there is no IPv6 support. Also see "[Using the CLI][CLI#ipv6-support]" for
how IPv6 can be enabled under Linux.
```

To stop the container:
```sh
docker stop zm-aio
```

## Invoking the command line tool using AIO image

The Zonemaster-CLI tool `zonemaster-cli` can also be invoked from the `zonemaster-all-in-one` image by adding the `cli` argument.

Moreover, there are two more usable Docker images to run `zonemaster-cli`:
`zonemaster/cli` and `zonemaster/backend`.
For more information on how to use the `zonemaster/backend` image to invoke
Zonemaster-CLI, see the [Docker Zonemaster-Backend] documentation.

```sh
docker run -t --rm zonemaster/all-in-one cli zonemaster.net --no-ipv6
```
or
```sh
docker run -t --rm zonemaster/all-in-one cli zonemaster.net
```

To make sure that Docker uses the latest version, add `--pull always`, e.g.

```sh
docker run -t --rm --pull always zonemaster/all-in-one zonemaster.net --no-ipv6
```

If `--pull always` is skipped, the invocation is quicker. The recommendation is
to include `--pull always` in the first command of a session to make sure that the
latest version is used, and then to exclude it to improve performance.

## Docker tips

If you run Docker on Mac computer with the M1 chip or if you want to have IPv6 inside

> Docker, see [Docker cli].

Reviewer: ```suggestion
If you run Docker on Mac computer with the M1 chip "[Using the CLI][CLI#ipv6-support]"
for more information.
```

[Get started]:                     https://www.docker.com/get-started/
[Docker Zonemaster-Backend]:       ../backend/Using-Zonemaster-Backend-Docker.md#invoking-the-command-line-tool-using-docker

> [Docker cli]:                      ../cli.md#docker-on-mac-with-m1-chip

Reviewer: ```suggestion
[CLI#docker-on-mac-with-m1-chip]:  ../cli.md#docker-on-mac-with-m1-chip
[CLI#ipv6-support]:                ../cli.md#ipv6-support
```

# Using the All-In-One docker image

## Table of contents
* [All-In-One](#all-in-one)
* [Using Docker](#using-docker)
* [Invoking the command line tool using AIO image](#invoking-the-command-line-tool-using-aio-image)
* [Docker tips](#docker-tips)

## All-In-One

> The `zonemaster-all-in-one` docker image provides an All-In-One solution containing Zonemaster-Backend, Zonemaster-CLI, and Zonemaster-GUI.

Reviewer 2: ```suggestion
The `zonemaster-all-in-one` docker image provides an All-In-One solution containing Zonemaster-CLI, Zonemaster-Backend, and Zonemaster-GUI.
```

> The `zonemaster-all-in-one` image can be run from computer that has Docker is installed.

Reviewer 2: ```suggestion
The `zonemaster-all-in-one` image can be run from any computer that has Docker installed.
```

### Using Docker

To run Zonemaster via the All-In-One Docker image you have to make sure that Docker is installed

> on the computer and that you can run Docker on it.

Reviewer 2: ```suggestion
To run Zonemaster via the All-In-One Docker image you have to make sure that Docker is installed
and running.
```

> * Instructions for installation are found on Docker [get started] page.

Reviewer 2: ```suggestion
* Instructions for installation are found on Docker [Get started] page.
```

* Run the command `docker ps` on the command line to verify that you can run
  Docker on the computer.

When Docker has been correctly installed, no more installation is needed to run
`zonemaster-cli`. Just follow the examples below.

There is a limitation in Docker regarding IPv6. Unless IPv6 has been enabled in
the Docker daemon, there is no support for IPv6. To avoid meaningless errors,
use `--no-ipv6` if there is no IPv6 support. Also see section "[IPv6 support]".

## Using the All-In-One image

> Start the All-In-One image with this command:

Reviewer 2: ```suggestion
Start the All-In-One container with this command:
```

```sh

> docker run --rm -d --name zmaio -p 8080:80 zonemaster/all-in-one gui

Reviewer 2: ```suggestion
docker run --rm -d --name zm-aio -p 8080:80 zonemaster/all-in-one gui
```

```

The Zonemaster GUI is now accessible via your browser at `http://localhost:8080/`.

> 

Reviewer: ```suggestion

When running on Docker on Windows IPv6 is not supported. Preferably select "IPv4 only" in GUI before starting a test.

```

Reviewer 2: Apparently there is, you need to change a configuration file though: see official documentation https://docs.docker.com/engine/daemon/ipv6/#use-ipv6-for-the-default-bridge-network and also e.g. https://github.com/nextcloud/all-in-one/blob/main/docker-ipv6-support.md#docker-desktop-windows-and-macos

> To stop the container

Reviewer 2: ```suggestion
To stop the container:
```

```sh

> docker stop zmaio

Reviewer 2: ```suggestion
docker stop zm-aio
```

```

> ## Invoking the command line tool using All-In-One image

Reviewer 2: ```suggestion
## Invoking the command line tool using AIO image
```

The zonemaster cli tool `zonemaster-cli` can also be invoked from the `zonemaster-all-in-one` image like 

> `zonemaster-cli` docker image, by adding `cli` argument.

Reviewer 2: ```suggestion
The Zonemaster-CLI tool `zonemaster-cli` can also be invoked from the `zonemaster-all-in-one` image by adding the `cli` argument.
```

There are two more usable Docker images to run `zonemaster-cli`:
`zonemaster/cli` and `zonemaster/backend`.
For more information on how to use `zonemaster/backend` image to invoke

> the cli, see the [Docker Zonemaster-Backend] documentation.

Reviewer 2: ```suggestion
Moreover, there are two more usable Docker images to run `zonemaster-cli`:
`zonemaster/cli` and `zonemaster/backend`.
For more information on how to use the `zonemaster/backend` image to invoke
Zonemaster-CLI, see the [Docker Zonemaster-Backend] documentation.
```

```sh
docker run -t --rm zonemaster/all-in-one cli zonemaster.net --no-ipv6
```
or
```sh
docker run -t --rm zonemaster/all-in-one cli zonemaster.net
```

To make sure that Docker uses the latest version, add `--pull always`, e.g.

```sh
docker run -t --rm --pull always zonemaster/all-in-one zonemaster.net --no-ipv6
```

If `--pull always` is skipped, the invocation is quicker. The recommendation is
to include `--pull always` in the first command of a session to make sure latest

> version is used, and then to exclude it to improve performance.

Reviewer 2: ```suggestion
to include `--pull always` in the first command of a session to make sure that the
latest version is used, and then to exclude it to improve performance.
```

## Docker tips

If you run Docker on Mac computer with the M1 chip or if you want to have ipv6 inside

> docker, see [Docker cli]

Reviewer 2: ```suggestion
If you run Docker on Mac computer with the M1 chip or if you want to have IPv6 inside
Docker, see [Docker cli].
```

[Get started]:                     https://www.docker.com/get-started/

> [Docker Zonemaster-Backend]:       backend/Using-Zonemaster-Backend-Docker.md#invoking-the-command-line-tool-using-docker

Reviewer 2: ```suggestion
[Docker Zonemaster-Backend]:       ../backend/Using-Zonemaster-Backend-Docker.md#invoking-the-command-line-tool-using-docker
```

> [Docker cli]:                      cli.md#docker-on-mac-with-m1-chip

Reviewer: ```suggestion
[Docker cli]:                      ../cli.md#docker-on-mac-with-m1-chip
```

* [All-In-One](#all-in-one)
* [Using Docker](#using-docker)
* [Invoking the command line tool using AIO image](#invoking-the-command-line-tool-using-aio-image)
* [Docker tips](#docker-tips)

## All-In-One

The `zonemaster-all-in-one` docker image provides an All-In-One solution containing Zonemaster-CLI, Zonemaster-Backend, and Zonemaster-GUI.

The `zonemaster-all-in-one` image can be run from any computer that has Docker installed.

### Using Docker

To run Zonemaster via the All-In-One Docker image you have to make sure that Docker is installed
and running.
* Instructions for installation are found on Docker [Get started] page.
* Run the command `docker ps` on the command line to verify that you can run
  Docker on the computer.

When Docker has been correctly installed, no more installation is needed to run
`zonemaster-cli`. Just follow the examples below.

Out of the box, Docker Desktop does not have support for IPv6 on neither Windows nor MacOS. We
have no clear information how IPv6 can be enabled. If you have not enabled IPv6 we suggest that
you run `zonemaster-cli` with `--no-ipv6` and in GUI select `IPv4 only` before starting a test
to avoid meaningless error.

> Also see "[Using the CLI][CLI#ipv6-support]" for how IPv6 can be enabled under Linux.

Reviewer 2: ```suggestion
Out of the box, Docker Desktop does not have official support for IPv6 on neither Windows nor MacOS. While it appears possible to enable it, to the best of our knowledge there is no official documentation for that. If you have not enabled IPv6 we suggest that you run `zonemaster-cli` with `--no-ipv6` and in GUI select `IPv4 only` before starting a test to avoid meaningless errors.
Also see "[Using the CLI][CLI#ipv6-support]" for how IPv6 can be enabled under Linux.
```

Reviewer: The suggestion from @tgreenx is also fine.

Reviewer: Maybe the following link should be added: https://docs.docker.com/engine/daemon/ipv6/

To stop the container:
```sh
docker stop zm-aio
```

## Invoking the command line tool using AIO image

The Zonemaster-CLI tool `zonemaster-cli` can also be invoked from the `zonemaster-all-in-one` image by adding the `cli` argument.

Moreover, there are two more usable Docker images to run `zonemaster-cli`:
`zonemaster/cli` and `zonemaster/backend`.
For more information on how to use the `zonemaster/backend` image to invoke
Zonemaster-CLI, see the [Docker Zonemaster-Backend] documentation.

```sh
docker run -t --rm zonemaster/all-in-one cli zonemaster.net --no-ipv6
```
or
```sh
docker run -t --rm zonemaster/all-in-one cli zonemaster.net
```

To make sure that Docker uses the latest version, add `--pull always`, e.g.

```sh
docker run -t --rm --pull always zonemaster/all-in-one zonemaster.net --no-ipv6
```

If `--pull always` is skipped, the invocation is quicker. The recommendation is
to include `--pull always` in the first command of a session to make sure that the
latest version is used, and then to exclude it to improve performance.

## Docker tips

If you run Docker on Mac computer with the M1 chip "[Using the CLI][CLI#ipv6-support]"

> for more information.

Reviewer: ```suggestion
If you run Docker on Mac computer with the M1 chip see 
"[Using the CLI][CLI#docker-on-mac-with-m1-chip]" for more information.
```

## Table of contents
* [All-In-One](#all-in-one)
* [Using Docker](#using-docker)
* [Invoking the command line tool using AIO image](#invoking-the-command-line-tool-using-aio-image)
* [Docker tips](#docker-tips)

## All-In-One

The `zonemaster-all-in-one` docker image provides an All-In-One solution containing Zonemaster-CLI, Zonemaster-Backend, and Zonemaster-GUI.

The `zonemaster-all-in-one` image can be run from any computer that has Docker installed.

### Using Docker

To run Zonemaster via the All-In-One Docker image you have to make sure that Docker is installed
and running.
* Instructions for installation are found on Docker [Get started] page.
* Run the command `docker ps` on the command line to verify that you can run
  Docker on the computer.

When Docker has been correctly installed, no more installation is needed to run
`zonemaster-cli`. Just follow the examples below.

There is a limitation in Docker regarding IPv6. Unless IPv6 has been enabled in
the Docker daemon, there is no support for IPv6. To avoid meaningless errors,
use `--no-ipv6` if there is no IPv6 support. Also see "[Using the CLI][CLI#ipv6-support]" for
how IPv6 can be enabled under Linux.

> 

Reviewer: I suggest that the following is added here:

"Out of the box,  Docker Desktop does not have support for IPv6 on neither Windows nor MacOS. We
have no clear information how IPv6 can be enabled. If you have not enabled IPv6 we suggest that
you run `zonemaster-cli` with `--no-ipv6` and in GUI select `IPv4 only` before starting a test"

I also suggest removing "When running on Docker on Windows (...) before starting a test." below to have the information in one place.

## Using the All-In-One image

Start the All-In-One container with this command:
```sh
docker run --rm -d --name zm-aio -p 8080:80 zonemaster/all-in-one gui
```

The Zonemaster GUI is now accessible via your browser at `http://localhost:8080/`.

> When running on Docker on Windows IPv6 is not supported. Preferably select "IPv4 only" in GUI before starting a test.

Reviewer 2: That's not exactly true though. See my [previous comment](https://github.com/zonemaster/zonemaster/pull/1487/changes#r3334265573).

Reviewer: It is quite unclear how it is. "IPv6 is only supported on Docker daemons running on Linux hosts" it says in https://docs.docker.com/engine/daemon/ipv6/

use `--no-ipv6` if there is no IPv6 support. Also see section "[IPv6 support]".

## Using the All-In-One image

Start the All-In-One image with this command:
```sh

> docker run --rm -d --name zmaio -p 8080:80 zonemaster/all-in-one gui

Reviewer: Is name "zmaio" something that you set and could be anything? You have to set name?

Author: It could be anything; giving the container a name makes it easier to stop.

Reviewer 2: `zm-aio` is probably nicer?
